extensions/ql-vscode/src/debugger/debug-session.ts

```diff
@@ -131,21 +128,129 @@ const QUERY_THREAD_ID = 1;
+ * having multiple properties within `QLDebugSession` that are only defined during query evaluation.
+ */
+class RunningQuery extends DisposableObject {
+  private readonly tokenSource = this.push(new CancellationTokenSource());
+  public readonly queryRun: CoreQueryRun;
+
+  public constructor(
+    queryRunner: QueryRunner,
+    config: CodeQLProtocol.LaunchConfig,
+    private readonly quickEvalContext: QuickEvalContext | undefined,
+    queryStorageDir: string,
+    private readonly logger: BaseLogger,
+    private readonly sendEvent: (event: Event) => void,
+  ) {
+    super();
+
+    // Create the query run, which will give us some information about the query even before the
+    // evaluation has completed.
+    this.queryRun = queryRunner.createQueryRun(
+      config.database,
+      {
+        queryPath: config.query,
+        quickEvalPosition: quickEvalContext?.quickEvalPosition,
+      },
+      true,
+      config.additionalPacks,
+      config.extensionPacks,
+      queryStorageDir,
+      undefined,
+      undefined,
+    );
+  }
+
+  public get id(): string {
+    return this.queryRun.id;
+  }
+
+  /**
+   * Evaluates the query, firing progress events along the way. The evaluation can be cancelled by
+   * calling `cancel()`.
+   *
+   * This function does not throw exceptions to report query evaluation failure. It just returns an
+   * evaluation result with a failure message instead.
+   */
+  public async evaluate(): Promise<
+    CodeQLProtocol.EvaluationCompletedEvent["body"]
+  > {
+    // Send the `EvaluationStarted` event first, to let the client known where the outputs are
+    // going to show up.
+    this.sendEvent(
+      new EvaluationStartedEvent(
+        this.queryRun.id,
+        this.queryRun.outputDir.querySaveDir,
+        this.quickEvalContext,
+      ),
+    );
+
+    try {
+      // Report progress via the debugger protocol.
+      const progressStart = new ProgressStartEvent(
+        this.queryRun.id,
+        "Running query",
+        undefined,
+        0,
+      );
+      progressStart.body.cancellable = true;
+      this.sendEvent(progressStart);
+      try {
+        return await this.queryRun.evaluate(
+          (p) => {
+            const progressUpdate = new ProgressUpdateEvent(
+              this.queryRun.id,
+              p.message,
+              (p.step * 100) / p.maxStep,
+            );
+            this.sendEvent(progressUpdate);
+          },
+          this.tokenSource.token,
+          this.logger,
+        );
+      } finally {
+        this.sendEvent(new ProgressEndEvent(this.queryRun.id));
+      }
+    } catch (e) {
+      const message = getErrorMessage(e);
+      return {
+        resultType: QueryResultType.OTHER_ERROR,
+        message,
+        evaluationTime: 0,
+      };
+    }
+  }
+
+  /**
+   * Attempts to cancel the running evaluation.
+   */
+  public cancel(): void {
+    this.tokenSource.cancel();
+  }
```

Comment:
Do you think we should be `cancel`ing the RunningQuery when it is disposed? It might be that the RunningQuery is disposed, but the query itself is still running. This can cause flakes in tests.

Reply:
See below